### ISSUE TITLE
Inclusive Inputs

### ISSUE THREAD

**Issue by nirjan-dev** [nirjan-dev/learning-to-code#279]
[link to the post](https://www.ovl.design/text/inclusive-inputs/)
# The basics
On first look, this structure does not look too bad. Semantic elements have been used. We have a label, which says that the user should enter her name. We have an input element. There is even an action on the form element. Did you know that you can submit forms without JavaScript? Old-school is calling.

```html
<form action="/contact" method="POST">
  <div class="form-group">
    <label>Name <span class="required">*</span></label>
    <input type="text" />
  </div>
</form>
```

}

But if we try to use that form with a screenreader, the information is sparse:
>>> edit text blank

Edit text blank. 59% of all input elements sound like this if you haven’t been blessed with eyesight. Edit text blank.

The reason for this output is that there is no link between the label and the input. The div it is wrapped in has no semantic meaning, neither does the CSS class. There are two ways to link a label with an element. 

# Association of labels and inputs
The first solution is to explicitly link the label and the input through an ID.

The input is – in spec terms – a labelable element. As long as it has an ID, it can be referenced by a label using the for attribute. As we explicitly say which input belongs to which label this combination is called explicit association.

```html
<div class="form-group">
  <label for="name">Name <span class="required">*</span></label>
  <input type="text" id="name" name="name" />
</div>
```
</div>

Alternatively we can put the input element inside of the label. This combination creates an implicit association. 

```html
<label for="name">
  Name <span class="required">*</span>
  <input type="text" id="name" name="name" />
</label>
```

Note: Some older screenreaders have a bug which requires the explicit link. Hence the for attribute is still set in this example. This might not be necessary if your users use up-to-date technology.

Both solutions produce the same output:

>>> name *, edit text with autofill menu
By associating labeland input we gave the latter an accessible name. The name attribute we added enables browsers to autofill the input with values a user has entered in fields with the same name before. Rather confusingly, though, name does not give an input its name. That’s the label’s job.

So far, so good. But you might notice the * inside the message. It often acts as a visual indicator that an input is required. The jury is out if should mark required fields or if the opposite approach – marking optional fields – might be more useful.

Either way, for screen readers the * is no useful information. Let’s fix this.

# marking required

The most obvious solution would be add the word required (or optional, if you want to go the opposite way) to the label text:

```html
<div>
  <label for="name">
    Name
    <span class="required">required</span>
  </label>
  <input type="text" id="name" name="name">
</div>
```

But, if the design can’t be changed, there are still ways to achieve add the necessary information.

To do so, we welcome ARIA in the mix and use an attribute called aria-hidden. It’s effect is in the name: Adding aria-hidden will hide the element and its children in the accessibility tree and thus from assistive technology. 

```html
<div class="form-group">
  <label for="name">
    Name
    <span class="required" aria-hidden="true">*</span>
    <span class="sr-only">required</span>
  </label>
  <input type="text" id="name" name="name">
</div>
```
We have added an additional span which contains the word required as well as the asterisk. By adding aria-hidden on the containing element it will not be announced by any assistive technology reading our page.

Scott O’Hara wrote an article about hiding content inclusively I refer time and time again too, if I hide things. If you want to take a deeper dive into this subject matter, it is recommended to head over there.

For sighted users it might look a bit weird to see «* required». This is solved by adding the sr-only class. Scott O’Hara proposed this implementation in the article linked above:

```css
.sr-only {
  clip-path: inset(100%);
  clip: rect(0 0 0 0);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}
```

These were two ways to add additional text describing in an input to a label. Both result in the same outcome:

>>> name required, edit text

At this point we have solved for some of the basic problems in marking up inputs. By associating a label element we made sure that the input has an accessible name. By adding textual alternatives for content we made sure that our users don’t have to be able to see an icon to recognise a required field. 

# Descriptions & Error Messages
The implementation so far is a fine baseline for more inclusive inputs. But some form elements will need to be described. And at one point there will be errors. Let’s take a look at how to solve this next.

# Descriptions

Given an input in which a user should add their name there probably is no explanation needed. But sometimes there is. Let’s take a password input as an example because almost every site has at least one requirement regarding this.

Informing users upfront is will prevent frustration and errors.

Let’s take a look at a standard way to do this, starting with what was discussed in the first part.

```html
<div class="form-group">
  <label for="password">
    Password
    <span class="required" aria-hidden="true">*</span>
    <span class="sr-only">required</span>
  </label>
  <input type="password" id="password" name="password">
  <p class="aside">Your password needs to have at least eight characters.</p>
</div>
```
I’ve added another element: the <p> containing a description, informing users about the format the password needs to have.

This is what VoiceOver has to say about this input:

>>> Password required, secure edit text
The description is lost. But we can recover it. To do so, we’ll need another way to link two elements. Unfortunately the for attribute only works between label and input elements and there is no native HTML for descriptions.

On the lucky side, though, ARIA has our back, again. This time the attribute to help out is aria-describedby. Simply put, it uses the text content on an element to describe another. As with all things accessibility there’s a lot more nuance than a single sentence to be taken into account. Scott O’Hara has written the definitive guide describing aria-describedby.

To solve the problem at hand the description needs to have an ID, which will in turn be referenced by the input. To update the example from above:

```html
<div class="form-group">
  <label for="password">
    Password
    <span class="required" aria-hidden="true">*</span>
    <span class="sr-only">required</span>
  </label>
  <input 
    type="password"
    id="password"
    name="password"
    aria-describedby="desc_pw"
  >
  <p class="aside" id="desc_pw">Your password needs to have at least eight characters.</p>
</div>
```
Adding these attributes, again, does not require any visual changes. Another error that can be fixed without anyone noticing.

>>> Password required, secure edit text

After announcing the basic information, there is a second set of output, containing complimentary information.

>>> Your password needs to have at least eight characters. You are currently on a text field, inside web content. To enter text in this field, type. This is a secure text field. Text in this field will not be displayed and it will not be spoken by VoiceOver

# Validation
With this we are at the last piece of the puzzle. Validation. Error messages are pretty much the same as the description we gave immediately. They will also be linked using aria-describedby. With one difference: Error messages are – mostly – not existing when the page is loading but a reaction to user input.

The trick here is that aria-describedby accepts a list of IDs which are concatenated to one spoken output.

To stay with the password example, I added an error message and added the ID of the description as well as the ID of the error message in aria-describedby:

```html
<div class="form-group">
  <label for="password">
    Password
    <span class="required" aria-hidden="true">*</span>
    <span class="sr-only">required</span>
  </label>
  <input 
    type="password"
    id="password"
    name="password"
    aria-describedby="error_pw desc_pw"
  >
  <p class="aside" id="desc_pw">Your password needs to have at least eight characters.</p>
  <p class="error" id="error_pw">Please check your input.</p>
</div>
```

If a user now focusses this input VoiceOver says the following:

>>> Password required, secure edit text. Please check your input. Your password need to have at least eight characters. You are currently on a text field, inside web content. To enter text in this field, type. This is a secure text field. Text in this field will not be displayed and it will not be spoken by VoiceOver

In this case the error message and the description complement each other. If an error message contains the text of the description this will lead to redundant output. In this case it might make sense to only describe the input with the error message.

Another thing to note is that there is no clear indication that the is invalid. Let’s fix this: 

```html
<div class="form-group">
  <label for="password">
    Password
    <span class="required" aria-hidden="true">*</span>
    <span class="sr-only">required</span>
  </label>
  <input 
    type="password"
    id="password"
    name="password"
    aria-invalid="true"
    aria-describedby="error_pw desc_pw"
  >
  <p class="aside" id="desc_pw">Your password needs to have at least eight characters.</p>
  <p class="error" id="error_pw">Please check your input.</p>
</div>
```

>>> Password required, invalid data, secure edit text.

# Staying accessible
Labelling and describing inputs is a low-hanging fruit. In most interfaces the elements are there, they just need to be structured properly. Still, it is virtually impossible to just make a website accessible. There is no just. Never. But you can work towards making it more accessible by the day (or sprint).

If you made it so far, you are probably already interested in fixing the web. And that’s amazing! Here are a few words of caution:

I often mentioned how many of these errors can be fixed without anyone noticing or the need to change the UI. While that’s certainly true, it can only solve so much.

Accessibility is a team-sport. If you want to make lasting change in your company, your dev team, basically anywhere you need to get the team onboard. There’s a limit on flying undercover and there is also a risk, because individual change makes it very hard to have a lasting impact.

Try to get the buy in of stakeholders, team members and leads. If your team has regular sessions in which you talk about tech, talk about accessibility. If not, why not start one?

Besides of educating your colleagues or community you’ll also need to somehow solidify the progress you make. If you work on a project with a Definition of Done, try to add accessibility testing to it. If you write tests, assert against ARIA states and properties.

And, most importantly, talk to your users or listen to those who do. Especially when working with assistive technology they are the experts and we as dev should be very humble and throw assumptions we might wide away.

Building user interfaces is a problem that is never finally solved. Staying accessible isn’t neither. It takes continuous work and iteration.
